src/lib/lore.sh

```diff
@@ -732,3 +735,93 @@ function save_new_lore_config()
 
   sed --in-place --regexp-extended "s<(${setting}=).*<\1${new_value}<" "$lore_config_path"
 }
+
+# TODO PR: Document
+# TODO PR: Test
+function apply_series()
+{
+  local series_url="$1"
+  local kernel_tree_path="$2"
+  local kernel_tree_branch="$3"
+  local flag="$4"
+  local path_to_tmp_dir
+  local series_path
+  local new_branch_name
+  local output
+  local cmd
+  local ret
+
+  if [[ -z "$kernel_tree_path" || -z "$kernel_tree_branch" ]]; then
+    complain 'The "Kernel Tree Path" and "Kernel Tree Branch" configs must be set.'
+    complain 'Go to Dashboard -> Settings.'
+    return 22 # EINVAL
+  fi
+
+  flag=${flag:-'SILENT'}
+
+  path_to_tmp_dir=$(mktemp --directory)
+
+  output=$(download_series "$series_url" "$path_to_tmp_dir")
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    complain "Download failed: ${output}"
+    return "$ret"
+  fi
+  series_path="$output"
+
+  cmd="git -C ${kernel_tree_path} switch ${kernel_tree_branch} --quiet"
+  cmd_manager "$flag" "$cmd"
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    complain "Could not switch kernel tree ${kernel_tree_path} to branch ${kernel_tree_branch}"
+    return "$ret"
+  fi
+
+  cmd="git -C ${kernel_tree_path} pull --quiet"
+  cmd_manager "$flag" "$cmd"
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    complain "Could not update branch ${kernel_tree_branch}"
+    return "$ret"
+  fi
+
+  new_branch_name=$(date +'%Y-%m-%d-%H-%M-%S-')
+  new_branch_name+=$(extract_message_id_from_url "$series_url")
+  cmd="git -C ${kernel_tree_path} checkout -b ${new_branch_name} --quiet"
+  cmd_manager "$flag" "$cmd"
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    complain "Could not create new branch ${new_branch_name} based on ${kernel_tree_branch}"
+    return "$ret"
+  fi
+
+  cmd="git -C ${kernel_tree_path} am --reject ${series_path} --quiet > /dev/null 2>&1"
+  cmd_manager "$flag" "$cmd"
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    if command_exists 'wiggle'; then
+      say 'Using wiggle to apply rejected patch.'
+      output=$(find "$kernel_tree_path" -type f -name '*.rej')
+      while IFS=$'\n' read -r rejected_file; do
+        original_file=$(printf '%s' "$rejected_file" | sed 's/\.rej//')
+        say "Applying rejected file '${rejected_file}' to '${original_file}'"
+        cmd_manager "$flag" "wiggle --replace ${original_file} ${rejected_file}"
+        ret="$?"
+        if [[ "$ret" != 0 ]]; then
+          complain 'Application of rejected file failed...'
+          return "$ret"
+        else
+          success 'Application of rejected file was a success!'
+        fi
+      done <<< "$output"
+      warning 'Application of rejected patch was a success!'
+      warning 'Recommended examining each replaced file for unresolved conflicts and semantic.'
+    else
+      complain "Could not apply series ${series_path}"
```

Comment:
I wonder if, after reaching this point,  kw should abort the `git am` operation it started, or even inform the user that they are in the middle of a `git am` and should themselves deal with this, i.e. abort, skip or solve the issue.
I do miss the error message, gone with `/dev/null 2>&1`, that would help understanding why the series could not be applied. I'm not sure, though, how this output could be presented to the user, as it doesn't seem suitable for the UI.

Reply:
Thanks for the review @magalilemes and nice catch!
You are right, we should abort or inform the user that there is a `git am` in progress. I opted for the latter thinking, as the application is encapsulated in a new branch, it is good to present the user with the option to deal with it but completely forgot to warn the user about it 😞. Because of the way dialog is implemented, any output to the stdout is not displayed correctly, which is why I used `/dev/null 2>&1` to not "pollute" the dialog's screen. Have the feeling that there is a better way to do this and I'm going to look at it.

Reply:
Good idea, going to work on that. I was thinking of not using
```bash
return "$ret"
```
throughout this function and returning static numbers for each error handling to make it predictable and possible to return a specific number here to signal the caller that a Yes/No screen needs to be created to prompt the user between these 3 options. What do you guys think?